lib/ansible/modules/cloud/amazon/ec2_group.py

```
@@ -284,6 +295,27 @@ def get_target_from_rule(module, ec2, rule, name, group, groups, vpc_id):
     return group_id, ip, target_group_created
 
 
+def ensure_tags(ec2, group, tags, changed, module):
+    try:
+        resource_id = group.id
+        cur_tags = group.tags
+        if cur_tags == tags:
```

If the new tags contain a value that is truthy (yes/true/etc) it will be interpreted as a bool so this will always show as changed since the cur_tags values will always be strings.

---

Hmm, should we explicitly convert the new tag values to strings, or is there a way to use the module parameter API to tell Ansible that we want a hash of strings and let it do the conversion for us?

---

I don't think there's a feature that will do this so yeah, I think we should probably iterate through the dict and use to_text to ensure all the values are strings. Instead of modifying with each value with 'str()' use 'to_text()' (you'll also need to add `from ansible.module_utils._text import to_text` to the imports). Besides that, this looks great. Thanks for the time you've spent on it :)

---

Would it be better to silently to_text values, or simply fail if they're not a string? If someone wants 'TRUE' and it gets converted to 'True', might it be better to fail with a message suggesting they quote the value in the YAML, as opposed to silently converting it? Other things like floats could be similarly impacted.

---

You're right about 'TRUE' being an issue. I don't know what to do about that. I'm lukewarm on failing since I can't think of any other cornercases. It shouldn't ever cause issues for integers or floats. Try creating a tag with the string "1" or "1.838" with your PR and if you check in the AWS console it rightly interprets them as numbers. But when boto is reading those numbers for us to compare they will always be strings...

---

The issue with floating point stuff is that it will drop trailing digits and/or lose precision. A tag value of 1.000 for example would be interpreted by Python as 1.0 unless properly quoted.

Either way - there's no consistent validation for tags across modules. I think I saw a PR float through recently that would add tag handling code to ec2.py, at which point all the modules should be converted over to use it for consistency's sake - but as of right now the issues we're talking about are widespread.

---

Good point!

---

https://github.com/ansible/ansible/pull/23387 is now merged, so you can use `compare_aws_tags` now